### DISCUSSION TITLE
expect(..) could return the value instead of void

### DISCUSSION THREAD

**Discussion by tarcisiofischer** [bitwizeshift/result#9] [Ideas]
[As in rust](https://doc.rust-lang.org/std/result/enum.Result.html#method.expect), would be nice if `expect(...)` would return the value instead of void, so that I could make things like

```
auto val = func().expect("Unexpected return value case");
```

AFAIK, the way of doing this today is to do this instead:

```
auto result = func();
assert(!result.has_error());
auto val = result.value();
```

**Implementation**

I didn't test, but I guess would be something like the code below. Inspired but the [current code](https://github.com/bitwizeshift/result/blob/master/include/result.hpp#L4715-L4741)

```
template <typename T, typename E>
template <typename String, typename>
inline RESULT_CPP14_CONSTEXPR
auto RESULT_NS_IMPL::result<T,E>::expect(String&& message)
  && -> typename std::add_lvalue_reference<T>::type
{
  if (has_error()) {
    detail::throw_bad_result_access_message(
      detail::forward<String>(message),
      static_cast<E&&>(m_storage.storage.m_error)
    );
  }
  return m_storage.storage.m_value;
}
```

Please let me know if this makes sense.

**Comment by bitwizeshift** [comment]
Hi @tarcisiofischer, thank you for the suggestion; this seems like a reasonable improvement. 

Funny thing is that I modeled a lot of `result`'s implementation after Rust, so it's weird that I didn't realize `expect` returns a reference to the object. 🤦‍♂️
